File: src/com/snake/and/ladder/Game.java
package com.snake.and.ladder;

import java.util.List;

public class Game {

  public void initialize(List<Players> players, List<Snakes> snakes, List<Ladders> ladders) {
    boolean isEndTheGame = false;
    int lastPosition;

    while (!isEndTheGame) {
      // person.forEach((i)-> System.out.println(i.getName()));
      int roll;

      for (Players player : players) {

        roll = Utility.roll();

        lastPosition = player.getPosition();
        player.setPosition(player.getPosition() + roll);

        // check if there is snake
        CheckSnakeForCurrentPosition(snakes, player);

        // check if there is a ladder
        CheckLadderForCurrentPosition(ladders,player);

        if (player.getPosition() >= 100) {
          System.out.println(player.getName() + " wins the game");
          isEndTheGame = true;
          break;
        }

        System.out.println(
            player.getName() + " rolled a " + roll + " and moved from " + lastPosition
                + " to " + player.getPosition());
      }
    }
  }

  private void CheckLadderForCurrentPosition(List<Ladders> ladders, Players player) {
    for (Ladders ladder : ladders) {
      if (ladder.getStart() == player.getPosition()) {
        player.setPosition(ladder.getEnd());
      }
    }
  }

  private void CheckSnakeForCurrentPosition(List<Snakes> snakes, Players player) {
    for (Snakes snake : snakes) {
      if (snake.getStart() == player.getPosition()) {
        player.setPosition(snake.getEnd());
      }
    }
  }

}

File: src/entities/Game.kt
package src.entities

import java.util.Queue

class Game(private val board: Board,
           private val players : Queue<Player>) {

    fun play() {

        val dice = board.dice
        while (players.size > 1) {
            val currPlayer = players.poll()
            val numberRolled = dice.rollDice()

            val currPosition = currPlayer.getPosition()
            val newPosition = calculateNewPosition(currPosition, numberRolled)

            println(currPlayer.name + " rolled a " + numberRolled + " and moved from " + currPosition + " to " + newPosition)

            if (newPosition == board.getSize()){
                println(currPlayer.name + " wins the game")
            }
            else{
                currPlayer.setPosition(newPosition)
                players.offer(currPlayer)
            }
        }
    }

    private fun calculateNewPosition(currPosition: Int, numberRolled: Int): Int {
        val newPos = currPosition + numberRolled
        if (newPos > board.getSize())
            return currPosition

        //Check if there's a snake at newPos
        board.snakeMap[newPos]?.let {
            return it.tail
        }

        //Check if there's a ladder at newPos
        board.ladderMap[newPos]?.let {
            return it.end
        }

        return newPos
    }
}
